File: packages/vscode-extension/src/background-agents.js
/**
 * Background-agents panel core — list the `cc agent --bg` supervisor sessions
 * straight from the state directory (`~/.chainlesschain/background-agents/`),
 * apply the same stale-heartbeat / dead-pid display correction the CLI's
 * `cc daemon status` performs (display-only: nothing is persisted from the
 * IDE), summarize, and tail logs. Session mutations (stop / rename / resume)
 * go through the CLI's `--json` commands; interactive attach goes through the
 * vendored agent-sdk pipe client — neither lives here.
 *
 * Pure Node (no `vscode`) → unit-testable; the webview glue lives in
 * ui/background-agents-view.js.
 */
const fs = require("fs");
const os = require("os");
const path = require("path");
const { spawnSync } = require("child_process");
const { isBlockingPhase } = require("./phase-attention");

const HEARTBEAT_STALE_MS = 120000;

// Pid identity (Gap 1: OS pid reuse — mirrors the CLI supervisor). A pid
// alone does not identify a process: after the worker dies the OS can hand
// the same pid to an unrelated process, so kill(pid, 0) alone would show a
// long-dead session as "running". Reuse is one-sided — a process that took
// over the pid can only have been created AFTER startedAt — so compare the
// pid's real creation time against the state file's startedAt.
const PID_IDENTITY_TOLERANCE_MS = 60000;
// The probe behind this cache is a SYNCHRONOUS wmic/powershell spawn
// (100-500ms on Windows) that blocks the extension host, and the Background
// Agents view polls the list every 3s — so the TTL directly sets how often
// the host stalls per running session. A process's creation time never
// changes while it lives; the only staleness risk is pid reuse (needs the
// worker to die first), which the display tolerates for one TTL window.
const START_TIME_CACHE_TTL_MS = 60000;
const LOG_TRUNCATION_NOTICE =
  "--- log truncated/rotated, resuming from tail ---";
const TRUNCATION_RESUME_TAIL_BYTES = 4096;

const startTimeCache = new Map();

function backgroundAgentsDir() {
  return path.join(os.homedir(), ".chainlesschain", "background-agents");
}

function isProcessAlive(pid, deps = {}) {
  const kill = deps.kill || process.kill.bind(process);
  if (!Number.isInteger(Number(pid)) || Number(pid) <= 0) return false;
  try {
    kill(Number(pid), 0);
    return true;
  } catch (error) {
    return error?.code === "EPERM";
  }
}

/** CIM_DATETIME (`20260711120000.500000+480`) → epoch ms, null when unparseable. */
function parseCimDateToMs(raw) {
  const m =
    /^(\d{4})(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})\.(\d{6})([+-])(\d{3})$/.exec(
      String(raw || "").trim(),
    );
  if (!m) return null;
  const localAsUtc = Date.UTC(
    Number(m[1]),
    Number(m[2]) - 1,
    Number(m[3]),
    Number(m[4]),
    Number(m[5]),
    Number(m[6]),
    Math.floor(Number(m[7]) / 1000),
  );
  const offsetMinutes = (m[8] === "-" ? -1 : 1) * Number(m[9]);
  return localAsUtc - offsetMinutes * 60000;
}

/**
 * One synchronous probe for a process's creation time (epoch ms); null when
 * it cannot be determined — callers FAIL OPEN to the plain kill(pid,0)
 * answer (this module is display-only, it never kills anything).
 */
function readProcessStartTimeMs(pid) {
  const target = Number(pid);
  if (!Number.isInteger(target) || target <= 0) return null;
  if (process.platform === "win32") {
    try {
      const r = spawnSync(
        "wmic",
        [
          "process",
          "where",
          `ProcessId=${target}`,
          "get",
          "CreationDate",
          "/value",
        ],
        { windowsHide: true, encoding: "utf8", timeout: 5000 },
      );
      if (!r.error && r.status === 0) {
        const m = /CreationDate=([^\r\n]+)/.exec(r.stdout || "");
        const ms = m ? parseCimDateToMs(m[1]) : null;
        if (ms !== null) return ms;
      }
    } catch (_err) {
      /* fall through to PowerShell */
    }
    try {
      const script =
        `$p = Get-CimInstance Win32_Process -Filter 'ProcessId=${target}' -ErrorAction SilentlyContinue; ` +
        "if ($p -and $p.CreationDate) { [DateTimeOffset]::new($p.CreationDate).ToUnixTimeMilliseconds() }";
      const r = spawnSync(
        "powershell",
        ["-NoProfile", "-NonInteractive", "-Command", script],
        { windowsHide: true, encoding: "utf8", timeout: 10000 },
      );
      if (!r.error && r.status === 0) {
        const n = Number(String(r.stdout || "").trim());
        if (Number.isFinite(n) && n > 0) return n;
      }
    } catch (_err) {
      /* fail open below */
    }
    return null;
  }
  try {
    const r = spawnSync("ps", ["-o", "lstart=", "-p", String(target)], {
      encoding: "utf8",
      timeout: 5000,
    });
    if (!r.error && r.status === 0) {
      const t = Date.parse(String(r.stdout || "").trim());
      if (Number.isFinite(t) && t > 0) return t;
    }
  } catch (_err) {
    /* fail open below */
  }
  return null;
}

/** Cached probe (default impl only — an injected deps probe runs uncached). */
function processStartTimeMs(pid, deps = {}) {
  if (typeof deps.readProcessStartTimeMs === "function") {
    const raw = deps.readProcessStartTimeMs(Number(pid));
    const n = Number(raw);
    return Number.isFinite(n) && n > 0 ? n : null;
  }
  const key = Number(pid);
  const at = Date.now();
  const hit = startTimeCache.get(key);
  if (hit && at - hit.at < START_TIME_CACHE_TTL_MS) return hit.value;
  const value = readProcessStartTimeMs(key);
  startTimeCache.set(key, { at, value });
  return value;
}

/**
 * Same-process check for a recorded pid. Fail-open rules match the CLI
 * supervisor: dead pid → false; no startedAt anchor → true (legacy states);
 * probe failure → true (never flip a live session to lost on a broken
 * probe); creation later than startedAt + tolerance → false (pid reused).
 */
function isSameProcess(pid, expectedStartedAtMs, deps = {}) {
  if (!isProcessAlive(pid, deps)) return false;
  const expected = Number(expectedStartedAtMs);
  if (!Number.isFinite(expected) || expected <= 0) return true;
  const actual = processStartTimeMs(pid, deps);
  if (actual === null) return true;
  return actual <= expected + PID_IDENTITY_TOLERANCE_MS;
}

/**
 * Display-only status correction, mirroring the CLI supervisor's
 * effectiveBackgroundAgentState: a "running" session whose heartbeat went
 * stale or whose worker pid is gone is SHOWN as lost — but the IDE never
 * writes the correction back (that's the CLI's job on its next touch).
 */
function effectiveStatus(state, { now = Date.now(), deps } = {}) {
  if (!state || state.status !== "running") {
    return {
      status: state?.status || "?",
      lostReason: state?.lostReason || null,
    };
  }
  if (
    Number.isFinite(Number(state.heartbeatAt)) &&
    now - Number(state.heartbeatAt) > HEARTBEAT_STALE_MS
  ) {
    return { status: "lost", lostReason: "heartbeat-stale" };
  }
  if (!isProcessAlive(state.pid, deps)) {
    return { status: "lost", lostReason: "process-exited" };
  }
  if (!isSameProcess(state.pid, state.startedAt, deps)) {
    // Gap 1: alive pid, but created well after this session started — the
    // OS reused the worker's pid for an unrelated process.
    return { status: "lost", lostReason: "pid-reused" };
  }
  return { status: "running", lostReason: null };
}

/**
 * List sessions from the state dir, newest first. Tolerant: unreadable or
 * malformed state files are skipped; a missing dir yields [].
 *
 * @returns {Array<{id,status,lostReason,phase,turnCount,title,cwd,sessionId,
 *   startedAt,endedAt,heartbeatAt,exitCode,logFile,interactive}>}
 */
function listBackgroundSessions({
  dir = backgroundAgentsDir(),
  now = Date.now(),
  deps,
} = {}) {
  let names;
  try {
    names = fs.readdirSync(dir);
  } catch {
    return [];
  }
  const sessions = [];
  for (const name of names) {
    if (!name.endsWith(".json") || name.includes(".job.")) continue;
    let state;
    try {
      state = JSON.parse(fs.readFileSync(path.join(dir, name), "utf8"));
    } catch {
      continue;
    }
    if (!state || typeof state.id !== "string") continue;
    const eff = effectiveStatus(state, { now, deps });
    const pendingApprovals = Number.isFinite(Number(state.pendingApprovals))
      ? Number(state.pendingApprovals)
      : 0;
    const pendingQuestion =
      state.pendingQuestion && typeof state.pendingQuestion === "object"
        ? {
            question:
              typeof state.pendingQuestion.question === "string"
                ? state.pendingQuestion.question
                : "",
            options: Array.isArray(state.pendingQuestion.options)
              ? state.pendingQuestion.options.filter(
                  (o) => typeof o === "string",
                )
              : [],
          }
        : null;
    sessions.push({
      id: state.id,
      status: eff.status,
      lostReason: eff.lostReason,
      phase: state.phase || null,
      pendingApprovals,
      // Blocked-on-human, computed once so every consumer (summary, sort,
      // badges, resume gating) classifies identically — no re-derived copies.
      attention:
        eff.status === "running" &&
        needsAttention(state.phase, pendingApprovals),
      pendingQuestion,
      uncertainSideEffects: Number.isFinite(Number(state.uncertainSideEffects))
        ? Number(state.uncertainSideEffects)
        : 0,
      turnCount: Number.isFinite(Number(state.turnCount))
        ? Number(state.turnCount)
        : null,
      title: typeof state.title === "string" ? state.title : "",
      cwd: state.cwd || "",
      sessionId: state.sessionId || null,
      startedAt: Number(state.startedAt) || null,
      endedAt: Number(state.endedAt) || null,
      heartbeatAt: Number(state.heartbeatAt) || null,
      exitCode: Number.isFinite(Number(state.exitCode))
        ? Number(state.exitCode)
        : null,
      logFile: state.logFile || path.join(dir, `${state.id}.log`),
      interactive: eff.status === "running" && Boolean(state.transport?.pipe),
    });
  }
  sessions.sort((a, b) => (b.startedAt || 0) - (a.startedAt || 0));
  return sessions;
}

/**
 * A session that is blocked on the human: the supervisor's phase reporter
 * writes `phase:"waiting_permission"` + `pendingApprovals` when the worker
 * parks on an approval, `needs_input` for a question block, and
 * `uncertain_side_effect` when a resumed turn must confirm irreversible ops.
 * Without this the session reads as "running fine" while it silently waits.
 * Delegates to the shared predicate so every surface classifies alike.
 */
function needsAttention(phase, pendingApprovals) {
  return isBlockingPhase(phase, pendingApprovals);
}

/** Status counts + running/interactive totals for the summary cards. */
function summarizeSessions(sessions) {
  const counts = {};
  let interactive = 0;
  let waiting = 0;
  for (const s of sessions || []) {
    counts[s.status] = (counts[s.status] || 0) + 1;
    if (s.interactive) interactive++;
    if (s.status === "running" && needsAttention(s.phase, s.pendingApprovals))
      waiting++;
  }
  return {
    total: (sessions || []).length,
    running: counts.running || 0,
    interactive,
    waiting,
    counts,
  };
}

/** Compact elapsed string ("42s" / "3m 12s" / "1h 4m"). */
function formatElapsed(session, now = Date.now()) {
  const end =
    Number.isFinite(Number(session?.endedAt)) && session.endedAt !== null
      ? Number(session.endedAt)
      : now;
  const start =
    Number.isFinite(Number(session?.startedAt)) && session.startedAt !== null
      ? Number(session.startedAt)
      : end;
  const ms = Math.max(0, end - start);
  const sec = Math.round(ms / 1000);
  if (sec < 60) return `${sec}s`;
  const min = Math.floor(sec / 60);
  if (min < 60) return `${min}m ${sec % 60}s`;
  return `${Math.floor(min / 60)}h ${min % 60}m`;
}

/** Last N lines of a session log ("" when unreadable/missing). */
function tailLog(logFile, lines = 120) {
  let text;
  try {
    text = fs.readFileSync(logFile, "utf8");
  } catch {
    return "";
  }
  const parts = text.split(/\r?\n/);
  const limit = Math.max(1, Number(lines) || 120);
  return parts.slice(-limit).join("\n");
}

/**
 * Read the log delta past `offset` (byte-oriented on the decoded string —
 * consistent as long as the same reader polls). Truncation/rotation (Gap 3):
 * instead of restarting from 0 — which replayed the ENTIRE new file into the
 * webview — emit an explicit marker and resume from the current tail.
 */
function readLogDelta(logFile, offset = 0) {
  let text;
  try {
    text = fs.readFileSync(logFile, "utf8");
  } catch {
    return { chunk: "", offset };
  }
  const from = offset;
  if (from > text.length) {
    const start = Math.max(0, text.length - TRUNCATION_RESUME_TAIL_BYTES);
    return {
      chunk: `\n${LOG_TRUNCATION_NOTICE}\n${text.slice(start)}`,
      offset: text.length,
      truncated: true,
    };
  }
  return { chunk: text.slice(from), offset: text.length };
}

module.exports = {
  HEARTBEAT_STALE_MS,
  LOG_TRUNCATION_NOTICE,
  PID_IDENTITY_TOLERANCE_MS,
  backgroundAgentsDir,
  effectiveStatus,
  isSameProcess,
  listBackgroundSessions,
  needsAttention,
  summarizeSessions,
  formatElapsed,
  tailLog,
  readLogDelta,
};

File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/BackgroundAgents.java
package com.chainlesschain.ide;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Background-agents panel core — list the {@code cc agent --bg} supervisor
 * sessions straight from the state directory
 * ({@code ~/.chainlesschain/background-agents/}), apply the same
 * stale-heartbeat display correction the CLI's {@code cc daemon status}
 * performs (display-only; the IDE never persists it), summarize, tail logs,
 * and render a text report for the dialog view. The Java twin of the VS Code
 * extension's background-agents.js. Pure (no IntelliJ SDK); the dialog glue
 * and the one-shot pipe client live elsewhere.
 */
public final class BackgroundAgents {

    /** Heartbeat staleness threshold, mirroring the CLI supervisor. */
    public static final long HEARTBEAT_STALE_MS = 120_000L;

    /**
     * Pid identity tolerance (VS twin PID_IDENTITY_TOLERANCE_MS): a pid alone
     * does not identify a process — after the worker dies the OS can hand the
     * same pid to an unrelated process. Reuse is one-sided (the impostor can
     * only have been created AFTER startedAt), so compare the pid's real
     * creation time against the state file's startedAt with this slack.
     */
    public static final long PID_IDENTITY_TOLERANCE_MS = 60_000L;

    /** How much of a log file the tail reader is willing to read (the old
     *  whole-file readString turned multi-MB logs into EDT-visible stalls). */
    static final int TAIL_READ_BYTES = 64 * 1024;

    private BackgroundAgents() {}

    /** One listed session (normalized, display-corrected). */
    public static final class Session {
        public final String id;
        public final String status;
        public final String lostReason;   // nullable
        public final String phase;        // nullable
        public final int turnCount;       // -1 when absent
        /** Approval requests blocking the worker (state pendingApprovals; 0 = none). */
        public final int pendingApprovals;
        public final String title;
        public final String cwd;
        public final String sessionId;    // nullable
        public final long startedAt;      // 0 when absent
        public final long endedAt;        // 0 when absent
        public final Integer exitCode;    // nullable
        public final String logFile;
        public final String pipePath;     // nullable — transport endpoint
        public final String token;        // nullable — transport auth
        public final boolean interactive;
        /** Unanswered ask_user_question parked by the worker (state pendingQuestion.question); nullable. */
        public final String pendingQuestion;
        /** Count of irreversible ops whose outcome is unknown after a resume (0 = none). */
        public final int uncertainSideEffects;

        Session(String id, String status, String lostReason, String phase, int turnCount,
                int pendingApprovals,
                String title, String cwd, String sessionId, long startedAt, long endedAt,
                Integer exitCode, String logFile, String pipePath, String token,
                boolean interactive, String pendingQuestion, int uncertainSideEffects) {
            this.id = id;
            this.status = status;
            this.lostReason = lostReason;
            this.phase = phase;
            this.turnCount = turnCount;
            this.pendingApprovals = pendingApprovals;
            this.title = title;
            this.cwd = cwd;
            this.sessionId = sessionId;
            this.startedAt = startedAt;
            this.endedAt = endedAt;
            this.exitCode = exitCode;
            this.logFile = logFile;
            this.pipePath = pipePath;
            this.token = token;
            this.interactive = interactive;
            this.pendingQuestion = pendingQuestion;
            this.uncertainSideEffects = uncertainSideEffects;
        }

        /** True when a human decision is blocking this session. */
        public boolean needsAttention() {
            return "running".equals(status)
                    && BackgroundAgents.needsAttention(phase, pendingApprovals);
        }
    }

    /** Default state directory ({@code ~/.chainlesschain/background-agents}). */
    public static Path defaultDir() {
        return Paths.get(System.getProperty("user.home"), ".chainlesschain", "background-agents");
    }

    /** One OS pid probe answer: liveness + creation time (null when unknown). */
    public static final class PidInfo {
        public final boolean alive;
        public final Long startMs; // nullable — creation time not determinable

        public PidInfo(boolean alive, Long startMs) {
            this.alive = alive;
            this.startMs = startMs;
        }
    }

    /** OS probe seam — injected in tests, {@link #probePid} in production. */
    public interface PidProbe {
        PidInfo probe(long pid);
    }

    /** Real OS probe via ProcessHandle. Kept tiny — the decision math is pure. */
    public static PidInfo probePid(long pid) {
        if (pid <= 0) return new PidInfo(false, null);
        try {
            java.util.Optional<ProcessHandle> h = ProcessHandle.of(pid);
            if (h.isEmpty() || !h.get().isAlive()) return new PidInfo(false, null);
            Long start = h.get().info().startInstant()
                    .map(java.time.Instant::toEpochMilli).orElse(null);
            return new PidInfo(true, start);
        } catch (Throwable t) {
            // probe failure = unknown → fail open (display-only correction)
            return new PidInfo(true, null);
        }
    }

    /**
     * Pure pid-identity decision (VS twin isSameProcess/effectiveStatus): the
     * lost-reason for a "running" session's pid probe answer, or null when the
     * pid still looks like the session's own worker. Unknown creation time or
     * unknown startedAt fails OPEN — this layer is display-only.
     */
    public static String pidLostReason(boolean alive, Long processStartMs, long stateStartedAtMs) {
        if (!alive) return "process-exited";
        if (stateStartedAtMs <= 0 || processStartMs == null) return null;
        return processStartMs <= stateStartedAtMs + PID_IDENTITY_TOLERANCE_MS
                ? null : "pid-reused";
    }

    /**
     * Display-only status correction: a "running" session whose heartbeat is
     * older than {@link #HEARTBEAT_STALE_MS} is shown as {@code lost}.
     * Returns {@code [status, lostReasonOrNull]}.
     */
    public static String[] effectiveStatus(String status, long heartbeatAt, long now) {
        return effectiveStatus(status, heartbeatAt, now, 0L, 0L, null);
    }

    /**
     * Full correction incl. pid identity (VS twin parity): after the heartbeat
     * check, a state pid that is gone → {@code lost/process-exited}; alive but
     * created well after startedAt → {@code lost/pid-reused} (the OS reused the
     * worker's pid). A state without a pid, or a null probe, skips the pid leg
     * (older CLI state files carry no pid — display-only, fail open).
     */
    public static String[] effectiveStatus(String status, long heartbeatAt, long now,
            long pid, long startedAt, PidProbe probe) {
        if (!"running".equals(status)) {
            return new String[] { status == null || status.isEmpty() ? "?" : status, null };
        }
        if (heartbeatAt > 0 && now - heartbeatAt > HEARTBEAT_STALE_MS) {
            return new String[] { "lost", "heartbeat-stale" };
        }
        if (pid > 0 && probe != null) {
            PidInfo info = probe.probe(pid);
            String reason = pidLostReason(info != null && info.alive,
                    info == null ? null : info.startMs, startedAt);
            if (reason != null) return new String[] { "lost", reason };
        }
        return new String[] { "running", null };
    }

    /**
     * List sessions in {@code dir}, newest first. Tolerant: unreadable or
     * malformed state files are skipped; a missing dir yields an empty list.
     * Uses the real OS pid probe; tests inject one via the overload.
     */
    public static List<Session> list(Path dir, long now) {
        return list(dir, now, BackgroundAgents::probePid);
    }

    /** {@link #list(Path, long)} with an injected pid probe (null = skip). */
    public static List<Session> list(Path dir, long now, PidProbe probe) {
        List<Session> out = new ArrayList<>();
        if (dir == null || !Files.isDirectory(dir)) return out;
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(dir, "*.json")) {
            for (Path file : stream) {
                String name = file.getFileName().toString();
                if (name.contains(".job.")) continue;
                Session s = readOne(file, dir, now, probe);
                if (s != null) out.add(s);
            }
        } catch (IOException ignored) {
            return out;
        }
        out.sort(Comparator.comparingLong((Session s) -> s.startedAt).reversed());
        return out;
    }

    private static Session readOne(Path file, Path dir, long now, PidProbe probe) {
        Map<String, Object> state;
        try {
            state = MiniJson.parseObject(Files.readString(file, StandardCharsets.UTF_8));
        } catch (Exception e) {
            return null;
        }
        if (state == null || !(state.get("id") instanceof String)) return null;
        String id = (String) state.get("id");
        long heartbeatAt = asLong(state.get("heartbeatAt"));
        long startedAt = asLong(state.get("startedAt"));
        String[] eff = effectiveStatus(asString(state.get("status")), heartbeatAt, now,
                asLong(state.get("pid")), startedAt, probe);
        String pipe = null;
        String token = null;
        Object transport = state.get("transport");
        if (transport instanceof Map) {
            Object p = ((Map<?, ?>) transport).get("pipe");
            Object t = ((Map<?, ?>) transport).get("token");
            if (p instanceof String) pipe = (String) p;
            if (t instanceof String) token = (String) t;
        }
        Object exit = state.get("exitCode");
        Integer exitCode = exit instanceof Number ? ((Number) exit).intValue() : null;
        String logFile = asString(state.get("logFile"));
        if (logFile == null || logFile.isEmpty()) {
            logFile = dir.resolve(id + ".log").toString();
        }
        boolean interactive = "running".equals(eff[0]) && pipe != null && !pipe.isEmpty();
        String pendingQuestion = null;
        Object pq = state.get("pendingQuestion");
        if (pq instanceof Map) {
            pendingQuestion = asString(((Map<?, ?>) pq).get("question"));
        }
        return new Session(
                id, eff[0], eff[1],
                asString(state.get("phase")),
                (int) asLong(state.get("turnCount"), -1L),
                (int) asLong(state.get("pendingApprovals"), 0L),
                asString(state.get("title"), ""),
                asString(state.get("cwd"), ""),
                asString(state.get("sessionId")),
                startedAt,
                asLong(state.get("endedAt")),
                exitCode, logFile, pipe, token, interactive,
                pendingQuestion,
                (int) asLong(state.get("uncertainSideEffects"), 0L));
    }

    /** Status → count rollup (insertion-ordered). */
    public static Map<String, Integer> summarize(List<Session> sessions) {
        Map<String, Integer> counts = new LinkedHashMap<>();
        for (Session s : sessions) counts.merge(s.status, 1, Integer::sum);
        return counts;
    }

    /** Compact elapsed string ("42s" / "3m 12s" / "1h 4m"). */
    public static String formatElapsed(long startedAt, long endedAt, long now) {
        long end = endedAt > 0 ? endedAt : now;
        long start = startedAt > 0 ? startedAt : end;
        long sec = Math.max(0, Math.round((end - start) / 1000.0));
        if (sec < 60) return sec + "s";
        long min = sec / 60;
        if (min < 60) return min + "m " + (sec % 60) + "s";
        return (min / 60) + "h " + (min % 60) + "m";
    }

    /**
     * True when a human decision is blocking the session: an approval that is
     * genuinely pending wins over any phase label; the canonical blocking
     * phases are {@code waiting_permission} / {@code needs_input} (kebab and
     * legacy {@code *_approval} spellings tolerated — mirrors the CLI's
     * background-agent-phase aliases).
     */
    public static boolean needsAttention(String phase, int pendingApprovals) {
        if (pendingApprovals > 0) return true;
        String p = phase == null ? "" : phase.trim().toLowerCase().replace('-', '_');
        return "waiting_permission".equals(p) || "needs_input".equals(p)
                || "uncertain_side_effect".equals(p)
                || p.contains("approval");
    }

    /**
     * Last {@code lines} lines of a log file ("" when unreadable). Reads only
     * the trailing {@link #TAIL_READ_BYTES} — long-lived agents grow multi-MB
     * logs, and the old whole-file readString made every refresh pay for all
     * of it. When the head was skipped, the first (possibly partial) line
     * after the seek point is dropped so a mid-UTF-8 landing never shows a
     * torn line.
     */
    public static String tailLog(String logFile, int lines) {
        String text;
        try {
            text = readTail(Paths.get(logFile), TAIL_READ_BYTES);
        } catch (Exception e) {
            return "";
        }
        String[] parts = text.split("\r?\n", -1);
        int limit = Math.max(1, lines);
        int from = Math.max(0, parts.length - limit);
        StringBuilder sb = new StringBuilder();
        for (int i = from; i < parts.length; i++) {
            if (i > from) sb.append('\n');
            sb.append(parts[i]);
        }
        return sb.toString();
    }

    /**
     * The trailing {@code capBytes} of a file, decoded as UTF-8. When the file
     * is larger than the cap, everything up to and including the first newline
     * in the window is dropped (the seek can land mid-line / mid-UTF-8-char);
     * a window without any newline is returned as-is rather than discarded.
     */
    static String readTail(Path file, int capBytes) throws IOException {
        try (java.nio.channels.SeekableByteChannel ch =
                Files.newByteChannel(file, java.nio.file.StandardOpenOption.READ)) {
            long size = ch.size();
            boolean skippedHead = size > capBytes;
            int want = (int) Math.min(size, capBytes);
            ch.position(size - want);
            java.nio.ByteBuffer bb = java.nio.ByteBuffer.allocate(want);
            while (bb.hasRemaining() && ch.read(bb) >= 0) {
                // keep filling — a channel may return short reads
            }
            byte[] bytes = new byte[bb.position()];
            bb.flip();
            bb.get(bytes);
            int from = 0;
            if (skippedHead) {
                for (int i = 0; i < bytes.length; i++) {
                    if (bytes[i] == (byte) '\n') {
                        from = i + 1;
                        break;
                    }
                }
            }
            return new String(bytes, from, bytes.length - from, StandardCharsets.UTF_8);
        }
    }

    /** One-line list row for the report / combo box. */
    public static String formatRow(Session s, long now) {
        StringBuilder sb = new StringBuilder();
        sb.append(String.format("%-9s", s.status));
        sb.append("  ").append(formatElapsed(s.startedAt, s.endedAt, now));
        sb.append("  ").append(s.title == null || s.title.isEmpty() ? s.id : s.title);
        if (s.phase != null) {
            sb.append("  (").append(s.phase);
            if (s.turnCount > 0) sb.append(" · turn ").append(s.turnCount);
            sb.append(')');
        }
        if (s.lostReason != null) sb.append("  [").append(s.lostReason).append(']');
        if ("running".equals(s.status) && needsAttention(s.phase, s.pendingApprovals)) {
            sb.append("  ⚠ ").append(attentionText(s));
        }
        if (s.interactive) sb.append("  ⇄");
        return sb.toString();
    }

    /**
     * Short blocking-state text, e.g. "waiting for approval (2 pending)",
     * "waiting for input: Deploy to prod?", or "confirm 3 uncertain side-effects".
     */
    private static String attentionText(Session s) {
        String p = s.phase == null ? "" : s.phase.trim().toLowerCase().replace('-', '_');
        if (s.pendingApprovals > 0) {
            return "waiting for approval (" + s.pendingApprovals + " pending)";
        }
        if ("uncertain_side_effect".equals(p)) {
            return s.uncertainSideEffects > 0
                    ? "confirm " + s.uncertainSideEffects + " uncertain side-effect"
                            + (s.uncertainSideEffects > 1 ? "s" : "")
                    : "confirm uncertain side-effects";
        }
        if ("needs_input".equals(p)) {
            if (s.pendingQuestion != null && !s.pendingQuestion.isEmpty()) {
                String q = s.pendingQuestion.length() > 140
                        ? s.pendingQuestion.substring(0, 140) : s.pendingQuestion;
                return "waiting for input: " + q;
            }
            return "waiting for input";
        }
        return "waiting for approval";
    }

    /** Multi-line detail block for one session (id/cwd/session/log tail). */
    public static String formatDetail(Session s, long now, int logLines) {
        StringBuilder sb = new StringBuilder();
        sb.append(s.id).append('\n');
        sb.append("status: ").append(s.status);
        if (s.lostReason != null) sb.append(" (").append(s.lostReason).append(')');
        if (s.exitCode != null) sb.append(" · exit ").append(s.exitCode);
        sb.append(" · ").append(formatElapsed(s.startedAt, s.endedAt, now)).append('\n');
        if (s.cwd != null && !s.cwd.isEmpty()) sb.append("cwd: ").append(s.cwd).append('\n');
        if (s.sessionId != null) sb.append("session: ").append(s.sessionId).append('\n');
        if ("running".equals(s.status) && needsAttention(s.phase, s.pendingApprovals)) {
            sb.append("attention: ").append(attentionText(s)).append('\n');
        }
        sb.append("interactive: ").append(s.interactive ? "yes ⇄" : "no").append('\n');
        String tail = tailLog(s.logFile, logLines);
        sb.append('\n').append(tail.isEmpty() ? "(no log output)" : tail);
        return sb.toString();
    }

    private static String asString(Object v) {
        return v instanceof String ? (String) v : null;
    }

    private static String asString(Object v, String fallback) {
        return v instanceof String ? (String) v : fallback;
    }

    private static long asLong(Object v) {
        return asLong(v, 0L);
    }

    private static long asLong(Object v, long fallback) {
        return v instanceof Number ? ((Number) v).longValue() : fallback;
    }
}
